### DISCUSSION TITLE
Maintain Temperature Feature

### DISCUSSION THREAD

**Discussion by bp529** [tagyoureit/nodejs-poolController#934] [Ideas]
My Jandy Aqualink board died and I made the switch to njsPC. Huge thank you @tagyoureit for maintaining this amazing project. I was able to quickly move over all existing equipment and integrate my IntelliflowVS over RS485 with ease.

Jandy had a really nice feature that I haven't figured out if it currently possible using Nixie or if external logic would be required.

When the heater is enabled for a body and Maintain Temperature is enabled (start and stop times) the pump will turn on at a specified interval for a few minutes to test the water temperature -- it will either heat to the set point if needed or turn off if no heat is needed. It then checks again at the next interval. 

I've searched through discussions and issues and haven't come across anything. 

Thank you!

**Comment by tagyoureit** [comment]
That's an interesting feature I haven't heard of, but curious as to it's actual use.  This sounds like it would be mostly useful at night when the main pump isn't running, but if that's the case than this necessarily means that the heat source would be a gas heater.  For the majority of users, using the gas heater for any prolonged period of time is prohibitively expensive.  

And I don't see a real use for this feature during the day because you can simply have a temperature set point that will be maintained.  

The closest feature that we have is that you can set pump speeds based on the solar or gas heater status.  Best practice with pool schedules is to have your lowest rpm/gpm setting be a speed that filters enough debris out of your pool on a daily basis and then if you need a speed boost for a cleaner/heater the app will dynamically boost the speed of the pump.  

Let me know if you are using this feature for another purpose than I've envisioned.

**Reply by bp529** [reply]
>  This sounds like it would be mostly useful at night when the main pump isn't running, but if that's the case than this necessarily means that the heat source would be a gas heater.

That is 100% correct. It saves electricity overnight if the pump doesn't need to be on (while my electric solar isn't producing) and keeps the pool at a desired temperature. I'll usually heat the pool for a 7 day period and drop the set point down while using maintain temp so I don't have to heat all the way back up to the desired set point when it's being used more

For example: 
Fri-Sun: 88* when most people are using the pool
Mon-Thur: 78* keeps the pool warm enough that it can be used at any time or set point can be raised higher and hit the new set point faster.

> For the majority of users, using the gas heater for any prolonged period of time is prohibitively expensive.

Unfortunately it has gotten pretty expensive. This feature helps alleviate some of the costs by not having to reheat the whole body every weekend. 

> The closest feature that we have is that you can set pump speeds based on the solar or gas heater status. Best practice with pool schedules is to have your lowest rpm/gpm setting be a speed that filters enough debris out of your pool on a daily basis and then if you need a speed boost for a cleaner/heater the app will dynamically boost the speed of the pump.

I'm currently doing this. Had the same setup with Jandy. Lower RPM for filtering, higher RPM for cleaning. 

Here is Jandy's documenation on the feature: [https://www.manualslib.com/manual/809789/Zodiac-Iaqualink.html?page=34](url)

I might create a fork and try my hand at implementing it. I appreciate your large commitment to this project and especially the open source.

**Reply by tagyoureit** [reply]
Adding this into the core code would be a pretty good sized effort.  Your easiest path would be to create a custom binding.  Here is one example I did a couple weeks ago ([valve custom bindings](https://github.com/tagyoureit/nodejs-poolController/discussions/922#discussioncomment-9027580)).

Essentially, I would set up schedules so the pool turns on every few hours for some period of time.  EG
M-Th:
 - 11pm-1am, Heat Mode: Heater, Set Point: 78
 - 3am-5am, Heat Mode: Heater, Set Point: 78

Fri-Sun:
 - 11pm-1am, Heat Mode: Heater, Set Point: 88
 - 3am-5am, Heat Mode: Heater, Set Point: 88
 
(Obviously, this is a complete guess so you can adjust to what works.)

The custom binding would very simply do a time check (>sunset and <sunrise) and also a check for set point > body temp.  If the conditions are true, turn off the pump.  This would cancel the remainder of the schedule.  Otherwise, it would stay on.  You can use an async timer to run every 30 minutes until the schedule is completed.

If want to look through the code and have a crack at it, this would be very doable.  If you want some help I can throw together some template code you can use.

**Reply by bp529** [reply]
I appreciate the detailed reply. It looks like custom bindings let you inject JS, very cool. I'll take a look at the code and choose what route to go. Thanks so much.
